Remove explicit JVM toolchain, add Java 8/Kotlin 2.0 compatibility

jamie10x opened the pull request:
This commit resolves two build configuration issues:

1. Removes the explicit `jvmToolchain(17)` declarations from the `timber` and `timber-lint` modules. By removing this, we allow the project to be built with the user's configured JDK (or the one used by Gradle), rather than forcing the use of JDK 17. This improves build flexibility for contributors using newer JDKs.

2. Corrects the DSL for setting the Android namespace. The build was failing with a `Could not find method namespace()` error due to using Groovy syntax (`namespace '...'`) with the Kotlin-based KMP Android plugin. This has been updated to the correct Kotlin syntax (`namespace = "..."`).

These changes complete the migration to the modern `com.android.kotlin.multiplatform.library` plugin and make the project easier to build for contributors.

Fixes #545

jamie10x committed:
Fix: Remove explicit JVM toolchain and correct KMP namespace syntax

This commit resolves two build configuration issues:

1. Removes the explicit `jvmToolchain(17)` declarations from the `timber` and `timber-lint` modules. By removing this, we allow the project to be built with the user's configured JDK (or the one used by Gradle), rather than forcing the use of JDK 17. This improves build flexibility for contributors using newer JDKs.

2. Corrects the DSL for setting the Android namespace. The build was failing with a `Could not find method namespace()` error due to using Groovy syntax (`namespace '...'`) with the Kotlin-based KMP Android plugin. This has been updated to the correct Kotlin syntax (`namespace = "..."`).

These changes complete the migration to the modern `com.android.kotlin.multiplatform.library` plugin and make the project easier to build for contributors.

Fixes #545

JakeWharton commented on timber-lint/build.gradle:
```diff
@@ -2,10 +2,6 @@ apply plugin: 'org.jetbrains.kotlin.jvm'
 apply plugin: 'com.android.lint'
 
-kotlin {
-  jvmToolchain(17)
```
We still need to target JDK 17 bytecode for this module

jamie10x replied:
I bring it back .

JakeWharton replied:
But the point of that issue is to remove toolchain usage. We should only be setting the source and target versions for the compilation.

jamie10x replied:
Understood, thanks for the clarification.

So the goal is to remove the strict toolchain requirement while ensuring the output bytecode is still compatible with Java 17.

My plan is to:
1. Remove `jvmToolchain(17)` from **both** `timber` and `timber-lint` build files.
2. In `timber-lint/build.gradle`, explicitly set `sourceCompatibility` and `targetCompatibility` to `JavaVersion.VERSION_17`.
3. In `timber/build.gradle`, add `compileOptions` to target Java 17 as well.

Does that align with what you are looking for? @JakeWharton

JakeWharton replied:
Yep, sounds about right.

JakeWharton commented on timber/build.gradle:
```diff
@@ -4,52 +4,49 @@ apply plugin: 'com.vanniktech.maven.publish'
-    optimization {
-      // "it" --> https://issuetracker.google.com/issues/445115242
-      it.consumerKeepRules.file('consumer-keep-rules.pro')
+    androidLibrary {
```
Please retain the two-space indent

jamie10x replied:
done

JakeWharton commented:
> due to using Groovy syntax (namespace '...') with the Kotlin-based KMP Android plugin. This has been updated to the correct Kotlin syntax (namespace = "...").

Just for reference, that's not what's happening here. Existing versions of Groovy support using the function call syntax for assignment, and that is being changed in newer versions. It has nothing to do with the language that the plugin is authored in.

jamie10x commented:
> > due to using Groovy syntax (namespace '...') with the Kotlin-based KMP Android plugin. This has been updated to the correct Kotlin syntax (namespace = "...").
> 
> Just for reference, that's not what's happening here. Existing versions of Groovy support using the function call syntax for assignment, and that is being changed in newer versions. It has nothing to do with the language that the plugin is authored in.

thank you for your feedback !

jamie10x committed:
Fix: Remove explicit JVM toolchain and correct KMP configuration

This commit resolves two build configuration issues:

1. Removes the explicit `jvmToolchain(17)` declarations. By using `sourceCompatibility` and `targetCompatibility` on the compile tasks instead, we ensure the output bytecode is compatible with Java 17 without forcing contributors to have that specific JDK installed. This improves build flexibility.

2. Corrects the DSL for setting the Android `namespace`. The build was failing due to an incorrect syntax attempt. This has been updated to the correct Kotlin DSL (`namespace = "..."`) inside the `androidLibrary` block.

These changes make the project easier to build for contributors and align with modern Gradle practices for KMP projects.

Fixes #545

JakeWharton committed:
Use plugin for setting compatibility

The library gets Java 8 and Kotlin 2.0 compatibility. The lint checks are Java 17 and we don't really need to care about the Kotlin compatibility.
